crates/cxx-qt-lib/src/gui/qimage.rs

```diff
@@ -332,6 +371,73 @@ impl QImage {
             None
         }
     }
+
+    /// Constructs a QImage from an existing memory buffer.
+    ///
+    /// # Safety
+    /// For details on safety see the [Qt documentation](https://doc.qt.io/qt-6/qimage.html#QImage-7)
+    pub unsafe fn from_raw_parts(
+        data: *const ffi::uchar,
+        width: i32,
+        height: i32,
+        format: QImageFormat,
+        image_cleanup_function: extern "C" fn(x: *mut ffi::c_void),
+        image_cleanup_data: *mut ffi::c_void,
+    ) -> Self {
+        ffi::qimage_init_from_raw_parts(
+            data,
+            width,
+            height,
+            format,
+            QImageCleanupFunction(image_cleanup_function),
+            image_cleanup_data,
+        )
+    }
+
+    /// Constructs a QImage from an existing mutable memory buffer.
+    ///
+    /// # Safety
+    /// For details on safety see the [Qt documentation](https://doc.qt.io/qt-6/qimage.html#QImage-8)
+    pub unsafe fn from_raw_parts_mut(
+        data: *mut ffi::uchar,
+        width: i32,
+        height: i32,
+        format: QImageFormat,
+        image_cleanup_function: extern "C" fn(x: *mut ffi::c_void),
+        image_cleanup_data: *mut ffi::c_void,
+    ) -> Self {
+        ffi::qimage_init_from_raw_parts_mut(
+            data,
+            width,
+            height,
+            format,
+            QImageCleanupFunction(image_cleanup_function),
+            image_cleanup_data,
+        )
+    }
+
+    /// Constructs a QImage from the raw data inside a `Vec<u8>`.
+    ///
+    /// # Safety
+    /// This function is unsafe because it requires that the data matches the given QImageFormat,
+    /// width and height.
+    ///
+    /// It is a convenience function around [Self::from_raw_parts_mut].
+    pub unsafe fn from_raw_bytes(
+        data: Vec<u8>,
```

Comment:
should this be a `&[u8]` as the Qt docs say it is read-only and in theory the lifetime out data should outlive the lifetime of the QImage.

But here we don't have any lifetime on QImage, similar to the QByteArray situation it raises the question if the Rust API should have two structs like a QImage and QImageRef in the future. Maybe we should raise an issue somewhere to discuss this and list the places where this could be useful?

Reply:
This function explictily takes a `Vec<u8>` so that we can pass the ownership of the data to the QImage, because we can't model the lifetime otherwise.

I also don't think it's really necessary to support a non-owning version of QImage, as the point of QImage is mostly to transfer ownership to Qt.
And as soon as we pass the FFI bridge, the lifetimes don't really mean much anyway.
So I'd rather just support this owning version then start to introduce lifetimes.

Also, side note: this version can actually modify the data, as it uses the non-const version of the constructor: https://doc.qt.io/qt-6/qimage.html#QImage-8 (i.e. `from_raw_parts_mut`).

Reply:
Ah i see so you are using the cleanup function to free it